include/quadrilateral_gimp_element.tcc

```diff
@@ -0,0 +1,178 @@
+// Return natural nodal coordinates
+template <>
+inline Eigen::MatrixXd
+    mpm::QuadrilateralGIMPElement<2, 16>::natural_nodal_coordinates() const {
+  //! Natural coordinates of nodes
+  Eigen::Matrix<double, 16, 2> local_nodes;
+  // clang-format off
+  (local_nodes << -1., -1.,
+                  1., -1.,
+                  1.,  1.,
+                 -1.,  1.,
+                 -3., -3.,
+                 -1., -3.,
+                  1., -3.,
+                  3., -3.,
+                  3., -1.,
+                  3.,  1.,
+                  3.,  3.,
+                  1.,  3.,
+                 -1.,  3.,
+                 -3.,  3.,
+                 -3.,  1.,
+                 -3., -1.).finished();
+  // clang-format on
+  return local_nodes;
+}
+
+//! Return shape functions of a 16-node Quadrilateral GIMP Element at a given
+//! local coordinate
+template <>
+inline Eigen::VectorXd mpm::QuadrilateralGIMPElement<2, 16>::shapefn(
+    const Eigen::Matrix<double, 2, 1>& xi, const VectorDim& particle_size,
+    const Eigen::Matrix<double, 2, 1>& deformation_gradient) const {
+
+  //! Dimension
+  const unsigned Dim = 2;
```

Comment:
Why do we need this?

Reply:
Remove this

Reply:
Tdim and Nfunctions are used in the for-loop but the shape function is not templatised with tdim and tnfunctions - so these are required

Reply:
Define it where you use it or just before